unblob/handlers/archive/hp/ipkg.py

```diff
@@ -0,0 +1,108 @@
+
+C_DEFINITIONS = r"""
+    typedef struct ipkg_file_entry {
+        char name[256];
+        uint64 offset;
+        uint64 size;
+        uint32 crc32;
+    } ipkg_toc_entry_t;
+
+    typedef struct ipkg_header {
+        char magic[4];
+        uint16 major;
+        uint16 minor;
+        uint32 toc_offset;
+        uint32 unknown_1;
+        uint32 toc_entries;
+        uint32 unknown_2[2];
+        uint32 always_null;
+        char file_version[256];
+        char product_name[256];
+        char ipkg_name[256];
+        char signature[256];
+    } ipkg_header_t;
+"""
+
+
+def is_valid_header(header: Instance) -> bool:
+    try:
+        snull(header.ipkg_name).decode("utf-8")
+        snull(header.file_version).decode("utf-8")
+        snull(header.product_name).decode("utf-8")
+    except UnicodeDecodeError:
+        return False
+    return True
+
+
+class HPIPKGExtractor(Extractor):
+    def __init__(self):
+        self._struct_parser = StructParser(C_DEFINITIONS)
+
+    def extract(self, inpath: Path, outdir: Path):
+        entries = []
+        with File.from_path(inpath) as file:
+            header = self._struct_parser.parse("ipkg_header_t", file, Endian.LITTLE)
+            file.seek(header.toc_offset, io.SEEK_SET)
+            for _ in range(header.toc_entries):
+                entry = self._struct_parser.parse(
+                    "ipkg_toc_entry_t", file, Endian.LITTLE
+                )
+                if not is_safe_path(outdir, Path(snull(entry.name).decode("utf-8"))):
+                    logger.warning(
+                        "Path traversal attempt, discarding.",
+                        outdir=outdir,
+                        outfile=Path(snull(entry.name).decode("utf-8")),
+                    )
+                    continue
+                entries.append(
+                    (
+                        outdir.joinpath(Path(snull(entry.name).decode("utf-8"))),
+                        Chunk(
+                            start_offset=entry.offset,
+                            end_offset=entry.offset + entry.size,
+                        ),
+                    )
+                )
+
+            for carve_path, chunk in entries:
+                carve_chunk_to_file(
+                    file=file,
+                    chunk=chunk,
+                    carve_path=carve_path,
+                )
+
+
+class HPIPKGHandler(StructHandler):
+    NAME = "ipkg"
+
+    PATTERNS = [HexString("69 70 6B 67 01 00 03 00")]
+
+    C_DEFINITIONS = C_DEFINITIONS
+    HEADER_STRUCT = "ipkg_header_t"
+    EXTRACTOR = HPIPKGExtractor()
+
+    def calculate_chunk(self, file: File, start_offset: int) -> Optional[ValidChunk]:
+        header = self.parse_header(file, endian=Endian.LITTLE)
+
+        if not is_valid_header(header):
+            raise InvalidInputFormat("Invalid IPKG header.")
+
+        file.seek(start_offset + header.toc_offset, io.SEEK_SET)
+        end_offset = -1
+        for _ in range(header.toc_entries):
```

Comment:
what if toc_entries == 0? we would have a strange end_offset?

Reply:
I do this check `if header.toc_offset == 0 or header.toc_entries == 0:` on the `is_valid_header`  function in the BDL handler, I can do it too on the ipkg handler